contracts/contracts/strategies/balancer/BaseBalancerStrategy.sol

```diff
@@ -0,0 +1,323 @@
+        uint256 yourPoolShare = IERC20(pTokenAddress)
+            .balanceOf(address(this))
+            .divPrecisely(IERC20(pTokenAddress).totalSupply());
+
+        uint256 balancesLength = balances.length;
+        for (uint256 i = 0; i < balancesLength; ++i) {
+            (address poolAsset, ) = toPoolAsset(_asset, 0);
+
+            if (address(tokens[i]) == poolAsset) {
+                (, uint256 assetAmount) = fromPoolAsset(
+                    poolAsset,
+                    balances[i].mulTruncate(yourPoolShare)
+                );
+                return assetAmount;
+            }
+        }
+    }
+
+    function getBPTExpected(address _asset, uint256 _amount)
+        internal
+        view
+        virtual
+        returns (uint256 bptExpected)
+    {
+        /* BPT price is calculated by dividing the pool (sometimes wrapped) market price by the
+         * rateProviderRate of that asset. To get BPT expected we need to multiply that by underlying
+         * asset amount divided by BPT token rate. BPT token rate is similar to Curve's virtual_price
+         * and expresses how much has the price of BPT appreciated in relation to the underlying assets.
+         *
+         * bptPrice = pool_asset_oracle_price / pool_asset_rate
+         *
+         * Since we only have oracle prices for the unwrapped version of the assets the equation
+         * turns into:
+         *
+         * bptPrice = from_pool_token(asset_amount).amount * oracle_price / pool_asset_rate
+         *
+         * bptExpected = bptPrice(in relation to specified asset) * asset_amount / BPT_token_rate
+         *
+         * and since from_pool_token(asset_amount).amount and pool_asset_rate cancel each-other out
+         * this makes the final equation:
+         *
+         * bptExpected = oracle_price * asset_amount / BPT_token_rate
+         *
+         * more explanation here:
+         * https://www.notion.so/originprotocol/Support-Balancer-OETH-strategy-9becdea132704e588782a919d7d471eb?pvs=4#382834f9815e46a7937f3acca0f637c5
+         */
+        address priceProvider = IVault(vaultAddress).priceProvider();
+        uint256 strategyAssetMarketPrice = IOracle(priceProvider).price(_asset);
+        uint256 bptRate = IRateProvider(platformAddress).getRate();
+
+        bptExpected = _amount
+            .mulTruncate(strategyAssetMarketPrice)
+            .divPrecisely(bptRate);
+    }
+
+    function getRateProviderRate(address _asset)
+        internal
+        view
+        virtual
+        returns (uint256);
+
+    function _lpDepositAll() internal virtual;
+
+    function _lpWithdraw(uint256 numBPTTokens) internal virtual;
+
+    function _lpWithdrawAll() internal virtual;
+
+    /**
+     * Balancer returns assets and rateProviders for corresponding assets ordered
+     * by numerical order.
+     */
+    function getPoolAssets() internal view returns (IERC20[] memory assets) {
+        (IERC20[] memory tokens, , ) = balancerVault.getPoolTokens(
```

Comment:
It's better to set directly to the returned `assets` variable rather than create a new variable `tokens` and then copy to the returned `assets` variable

Reply:
good point 👍